File: 79.word-search.js
/*
 * @lc app=leetcode id=79 lang=javascript
 *
 * [79] Word Search
 */

// @lc code=start
/**
 * @param {character[][]} board
 * @param {string} word
 * @return {boolean}
 */
var exist = function(board, word) {
  // Brute-force backtracking
  // O(n*m* 4^n)
  // DFS implementation
  let ROWS = board.length, COLS = board[0].length
  let path = []
  let wordIndex = 0
  let dfs = (r, c, wordIndex) => {
    if (wordIndex === word.length) return true
    if (r < 0 || c < 0 || r >= ROWS || c >= COLS ||
        path.some(v => v[0] === r && v[1] === c) ||
        word[wordIndex] !== board[r][c]
       ) {
          return false
        }
    path.push([r, c])
    let res = (
      dfs(r + 1, c, wordIndex + 1) ||
      dfs(r - 1, c, wordIndex + 1) ||
      dfs(r, c + 1, wordIndex + 1) ||
      dfs(r, c - 1, wordIndex + 1)
    )
    path.pop()
    return res
  }

  for (let i = 0; i < board.length; i++) {
    for (let j = 0; j < board[i].length; j++) {
      if (dfs(i, j, wordIndex)) return true
    }
  }
  return false
};
// @lc code=end

let board = [
  ["A","B","C","E"],
  ["S","F","C","S"],
  ["A","D","E","E"]
]

console.log(
  exist(
    board,
    "ABCCED"
  )
)

File: cpp/79.word-search.cpp
/*
 * @lc app=leetcode id=79 lang=cpp
 *
 * [79] Word Search
 */

// @lc code=start
class Solution
{
public:
    bool dfsHelper(vector<vector<char>> &board, int row, int col, string &word, int pos)
    {
        // base cases
        if (pos >= word.size())
        {
            return true;
        }
        // bounds checks
        if (row >= board.size() || row < 0)
        {
            return false;
        }
        if (col >= board[0].size() || col < 0)
        {
            return false;
        }
        // check if board index matches char at wordPos
        if (board[row][col] != word[pos])
        {
            return false;
        }
        // mark positions on board to prevent backtracks
        char ch = board[row][col];
        board[row][col] = '*';

        // recursive subcalls
        int up = row + 1;
        int down = row - 1;
        int left = col - 1;
        int right = col + 1;

        pos++;
        // keep moving up, down, left, right
        bool res = (dfsHelper(board, up, col, word, pos) ||
                    dfsHelper(board, down, col, word, pos) ||
                    dfsHelper(board, row, left, word, pos) ||
                    dfsHelper(board, row, right, word, pos));

        // unmark board positions
        board[row][col] = ch;

        return res;
    }
    bool exist(vector<vector<char>> &board, string word)
    {
        // DFS recursive approach
        // iterate over every board position
        // launch DFS with word, moving up/down or left/right
        // O((n*m)^(n*m)) time O((n*m)^(n*m)) space
        for (int i = 0; i < board.size(); i++)
        {
            for (int j = 0; j < board[0].size(); j++)
            {
                if (dfsHelper(board, i, j, word, 0))
                {
                    return true;
                }
            }
        }
        return false;
    }
};
// @lc code=end
